MSBuild Error: Failed posting a file copy to an ActionBlock. Should not happen with block at max int capacity

An interesting one, gotten from hitting Ctrl-C during a build of dotnet/runtime using 6.0.100-rc.2.21505.57

This line
https://github.com/dotnet/msbuild/blob/a59d7a533c9154e8aa99b823625e7eff199ddf1a/src/Tasks/Copy.cs#L588
should perhaps be
```c#
                if (!partitionAccepted && !CancellationToken.IsCancellationRequested)
```
```
Build FAILED.

C:\git\runtime\.dotnet\sdk\6.0.100-rc.2.21505.57\Microsoft.Common.CurrentVersion.targets(4812,5): error MSB4018: The "Copy" task failed unexpectedly. [C:\git\runtime\src\libraries\System.IO.FileSystem\tests\ManualTests\System.IO.FileSystem.Manual.Tests.csproj]
C:\git\runtime\.dotnet\sdk\6.0.100-rc.2.21505.57\Microsoft.Common.CurrentVersion.targets(4812,5): error MSB4018: Microsoft.Build.Shared.InternalErrorException: MSB0001: Internal MSBuild Error: Failed posting a file copy to an ActionBlock. Should not happen with block at max int capacity. [C:\git\runtime\src\libraries\System.IO.FileSystem\tests\ManualTests\System.IO.FileSystem.Manual.Tests.csproj]
C:\git\runtime\.dotnet\sdk\6.0.100-rc.2.21505.57\Microsoft.Common.CurrentVersion.targets(4812,5): error MSB4018:    at Microsoft.Build.Tasks.Copy.CopyParallel(CopyFileWithState copyFile, Int32 parallelism, List`1& destinationFilesSuccessfullyCopied) [C:\git\runtime\src\libraries\System.IO.FileSystem\tests\ManualTests\System.IO.FileSystem.Manual.Tests.csproj]
C:\git\runtime\.dotnet\sdk\6.0.100-rc.2.21505.57\Microsoft.Common.CurrentVersion.targets(4812,5): error MSB4018:    at Microsoft.Build.Tasks.Copy.Execute(CopyFileWithState copyFile, Int32 parallelism) [C:\git\runtime\src\libraries\System.IO.FileSystem\tests\ManualTests\System.IO.FileSystem.Manual.Tests.csproj]
C:\git\runtime\.dotnet\sdk\6.0.100-rc.2.21505.57\Microsoft.Common.CurrentVersion.targets(4812,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() [C:\git\runtime\src\libraries\System.IO.FileSystem\tests\ManualTests\System.IO.FileSystem.Manual.Tests.csproj]
C:\git\runtime\.dotnet\sdk\6.0.100-rc.2.21505.57\Microsoft.Common.CurrentVersion.targets(4812,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask) [C:\git\runtime\src\libraries\System.IO.FileSystem\tests\ManualTests\System.IO.FileSystem.Manual.Tests.csproj]
C:\Users\danmose\.nuget\packages\microsoft.net.compilers.toolset\4.0.0-5.21453.15\tasks\netcoreapp3.1\Microsoft.CSharp.Core.targets(75,5): error MSB5021: Terminating the task executable "dotnet" and its child processes because the build was canceled. [C:\git\runtime\src\libraries\Microsoft.Extensions.Logging.Abstractions\tests\Microsoft.Extensions.Logging.Generators.Tests\Microsoft.Extensions.Logging.Generators.Roslyn4.0.Tests.csproj]
C:\Users\danmose\.nuget\packages\microsoft.net.compilers.toolset\4.0.0-5.21453.15\tasks\netcoreapp3.1\Microsoft.CSharp.Core.targets(75,5): error MSB4181: The "Csc" task returned false but did not log an error. [C:\git\runtime\src\libraries\Microsoft.Extensions.Logging.Abstractions\tests\Microsoft.Extensions.Logging.Generators.Tests\Microsoft.Extensions.Logging.Generators.Roslyn4.0.Tests.csproj]
    0 Warning(s)
    3 Error(s)

```

--- a/src/Tasks/Copy.cs
+++ b/src/Tasks/Copy.cs
@@ -586,11 +586,18 @@ namespace Microsoft.Build.Tasks
 
             foreach (List<int> partition in partitionsByDestination.Values)
             {
-                bool partitionAccepted = partitionCopyActionBlock.Post(partition);
-                if (!partitionAccepted)
+                if (_cancellationTokenSource.IsCancellationRequested)
                 {
-                    // Retail assert...
-                    ErrorUtilities.ThrowInternalError("Failed posting a file copy to an ActionBlock. Should not happen with block at max int capacity.");
+                    break;
+                }
+                else
+                {
+                    bool partitionAccepted = partitionCopyActionBlock.Post(partition);
+                    if (!partitionAccepted)
+                    {
+                        // Retail assert...
+                        ErrorUtilities.ThrowInternalError("Failed posting a file copy to an ActionBlock. Should not happen with block at max int capacity.");
+                    }
                 }
             }
 

--- a/src/Tasks/Copy.cs
+++ b/src/Tasks/Copy.cs
@@ -596,7 +596,6 @@ namespace Microsoft.Build.Tasks
                     // Retail assert...
                     ErrorUtilities.ThrowInternalError("Failed posting a file copy to an ActionBlock. Should not happen with block at max int capacity.");
                 }
-
             }
 
             partitionCopyActionBlock.Complete();

--- a/src/Tasks/Copy.cs
+++ b/src/Tasks/Copy.cs
@@ -304,7 +304,7 @@ namespace Microsoft.Build.Tasks
 
                 File.Copy(sourceFileState.Name, destinationFileState.Name, true);
             }
-            
+
             // Files were successfully copied or linked. Those are equivalent here.
             WroteAtLeastOneFile = true;
 
@@ -586,19 +586,17 @@ namespace Microsoft.Build.Tasks
 
             foreach (List<int> partition in partitionsByDestination.Values)
             {
+                bool partitionAccepted = partitionCopyActionBlock.Post(partition);
                 if (_cancellationTokenSource.IsCancellationRequested)
                 {
                     break;
                 }
-                else
+                else if (!partitionAccepted)
                 {
-                    bool partitionAccepted = partitionCopyActionBlock.Post(partition);
-                    if (!partitionAccepted)
-                    {
-                        // Retail assert...
-                        ErrorUtilities.ThrowInternalError("Failed posting a file copy to an ActionBlock. Should not happen with block at max int capacity.");
-                    }
+                    // Retail assert...
+                    ErrorUtilities.ThrowInternalError("Failed posting a file copy to an ActionBlock. Should not happen with block at max int capacity.");
                 }
+
             }
 
             partitionCopyActionBlock.Complete();
@@ -948,17 +946,17 @@ namespace Microsoft.Build.Tasks
             return String.Equals(fullSourcePath, fullDestinationPath, filenameComparison);
         }
 
-    	private static int GetParallelismFromEnvironment()
-	    {
-	        int parallelism = Traits.Instance.CopyTaskParallelism;
-	        if (parallelism < 0)
-	        {
-	            parallelism = DefaultCopyParallelism;
-	        }
+        private static int GetParallelismFromEnvironment()
+        {
+            int parallelism = Traits.Instance.CopyTaskParallelism;
+            if (parallelism < 0)
+            {
+                parallelism = DefaultCopyParallelism;
+            }
             else if (parallelism == 0)
-	        {
-	            parallelism = int.MaxValue;
-	        }
+            {
+                parallelism = int.MaxValue;
+            }
             return parallelism;
         }
     }

